lib/core/text/u16_string.nit

```diff
@@ -0,0 +1,125 @@
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Wrapper of ICU's UTF-16 encoded strings and conversion
+# This module is meant to ease the use of complex string operations provided by the ICU library.
+# The module provides a wrapper for ICU's string structure : `UChar *` as well as conversion fucntions to/from `String` and `CString`
+module u16_string is pkgconfig ("icu-io", "icu-i18n", "icu-uc")
+
+`{
+	#include <unicode/utypes.h>
+	#include <unicode/ustring.h>
+`}
+
+# UTF-16 encoded string
+class U16String
+	super Finalizable
+
+	# Pointer to a `UChar *` string
+	var uchar_string : UCharString is noinit
+
+	# Number of characters in `uchar_string`
+	var length = 0
+
+	# Number of bytes in `uchar_string`
+	var bytes = 0
+
+	# Returns a converted `UCharString` from a `String`
+	init (source : String)
+	do
+		var csource = source.to_cstring
+		var csource_length = source.to_cstring.cstring_length
+
+		uchar_string = new UCharString.nul
+		var required_length = uchar_string.from_cstring(0, csource, csource_length)
+
+		uchar_string = new UCharString.empty(required_length + 1)
+		uchar_string.from_cstring(required_length + 1, csource, csource_length)
+
+		length = csource.to_s.length
+		bytes = csource_length
+	end
+
+	# Returns an empty `U16String` of length `len`
+	init empty (len : Int) do
+		uchar_string = new UCharString.empty(len)
+		length = len - 1
+	end
+
+	# Returns a null `U16String`
+	init nul do uchar_string = new UCharString.nul
+
+	# Get char at `index`
+	fun [](index: Int): Char do return uchar_string[index]
```

Comment:
Asking myself if this could subclass `Sequence[Char]` or something similar.

Reply:
Same as the `[]=`, this won't work on characters >0xFFFF

Reply:
What about returning a `U16Char`?

Reply:
Historically, `String` was a subclass of `SequenceRead[Char]` but this caused some issues especially because a lot of methods of `Collection` & cie was inherited and caused some noise and confusion in what is string-related and what was collection-related.

Another issue was that at the time is was not clear if String is a sequence of byte of a sequence of code points (or a sequence of glyph, etc.).
To solve the issue, two view where attached to the String: `chars` a code point view (as a `SequenceRead[Char]`) and `bytes` a byte view (as a `SequenceRead[Int]`). #260